src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -31,49 +32,18 @@ public static class ActivatorUtilities
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
             params object[] parameters)
         {
-            int bestLength = -1;
-            bool seenPreferred = false;
-
-            ConstructorMatcher bestMatcher = default;
-
-            if (!instanceType.IsAbstract)
-            {
-                foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
-                {
-                    var matcher = new ConstructorMatcher(constructor);
-                    bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
-
-                    if (isPreferred)
-                    {
-                        if (seenPreferred)
-                        {
-                            ThrowMultipleCtorsMarkedWithAttributeException();
-                        }
-
-                        if (length == -1)
-                        {
-                            ThrowMarkedCtorDoesNotTakeAllProvidedArguments();
-                        }
-                    }
-
-                    if (isPreferred || bestLength < length)
-                    {
-                        bestLength = length;
-                        bestMatcher = matcher;
-                    }
-
-                    seenPreferred |= isPreferred;
-                }
-            }
-
-            if (bestLength == -1)
-            {
-                string? message = $"A suitable constructor for type '{instanceType}' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.";
-                throw new InvalidOperationException(message);
-            }
-
-            return bestMatcher.CreateInstance(provider);
+            Type[] argumentTypes = parameters
```

Comment:
To avoid LINQ usage, you could replace `Select` and `ToArray` with `Array.ConvertAll`. For least allocations, you could use a manual `for` loop.

Reply:
Changed this to use a simple `for` loop. 👍